desc/integrals/basis.py

```diff
@@ -58,6 +85,627 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
     return put(is_intersect, idx[0], edge_case)
 
 
+def _chebcast(cheb, arr):
+    """Add leftmost axis to ``cheb`` depending on ``arr.ndim``.
+
+    Input ``arr`` should not have rightmost dimension of cheb that iterates
+    coefficients, but may have additional leftmost dimension for batch operation.
+    """
+    errorif(
+        jnp.ndim(arr) > cheb.ndim,
+        NotImplementedError,
+        msg=f"Only one additional axis for batch dimension is allowed. "
+        f"Got {jnp.ndim(arr) - cheb.ndim + 1} additional axes.",
+    )
+    return cheb if jnp.ndim(arr) < cheb.ndim else cheb[jnp.newaxis]
+
+
+class FourierChebyshevSeries(IOAble):
+    """Fourier-Chebyshev series.
+
+    f(x, y) = ∑ₘₙ aₘₙ ψₘ(x) Tₙ(y)
+    where ψₘ are trigonometric polynomials on [0, 2π]
+    and Tₙ are Chebyshev polynomials on [−yₘᵢₙ, yₘₐₓ].
+
+    Notes
+    -----
+    Performance may improve significantly
+    if the spectral resolutions ``M`` and ``N`` are powers of two.
```

Comment:
usually for fourier you need an odd number of points?

Reply:
i forget the exact reason, but usually when solving PDEs you want to use an odd number of points for fourier discretizations. Intuitively you can imagine that with an even number of points, in the spectral domain you have a sin(nx) but no cos(nx) or something?

Reply:
> even number of points, in the spectral domain you have a sin(nx) but no cos(nx) or something?
Right, but you should account for that when you evaluate the fourier series. FFT algorithms use divide and conqur algorithms that perform better with powers of two